src/main/java/com/google/jenkins/plugins/storage/StorageUtil.java

```diff
@@ -0,0 +1,183 @@
+   * @param workspace The workspace containing the included file.
+   * @return The unix-style relative path of file.
+   * @throws UploadException when the input is malformed
+   */
+  public static String getRelative(FilePath include, FilePath workspace)
+      throws UploadException {
+    LinkedList<String> segments = new LinkedList<String>();
+    while (!include.equals(workspace)) {
+      segments.push(include.getName());
+      include = include.getParent();
+      if (Strings.isNullOrEmpty(include.getName())) {
+        // When we reach "/" we're done either way.
+        break;
+      }
+    }
+    return Joiner.on("/").join(segments);
+  }
+
+  /**
+   * If a path prefix to strip has been specified, and the input string
+   * starts with that prefix, returns the portion of the input after that
+   * prefix. Otherwise, returns the unmodified input.
+   */
+  protected static String getStrippedFilename(String filename, String pathPrefix) {
+    if (pathPrefix != null && filename != null
+        && filename.startsWith(pathPrefix)) {
+      return filename.substring(pathPrefix.length());
+    }
+    return filename;
+  }
+
+  /**
+   * Perform variable expansion for non-pipeline steps.
+   *
+   * @param name The name, potentially including with variables
+   * @param run The current run, used to determine pipeline status and to get environment.
+   * @param listener Task listener, used to get environment
+   * @return The updated name, with variables resolved
+   */
+  public static String replaceMacro(String name, Run<?,?> run, TaskListener listener)
+      throws InterruptedException, IOException {
+    if(run instanceof AbstractBuild) {
+      name = Util.replaceMacro(name, run.getEnvironment(listener));
+    }
+    return name;
+  }
+
+  public static class BucketPath implements Serializable {
+    /**
+     * Prepares a new BucketPath.
+     * @param uri path to the bucket object, of the form "gs://bucket_name/path/to/object". May
+     *            contain unresolved variables, which will get resolved for non-pipeline cases.
+     *            May contain other characters (i.e., *), no verification is done in this class.
+     * @param run The currect run, used for resolving variables
+     * @param listener Task listener, used for resolving variables
+     * @param moduleName Name of the throwing module, to be included in error messages.
+     */
+    public BucketPath(String uri, Run<?, ?> run, TaskListener listener, String moduleName) throws IOException, InterruptedException {
+      this.original = uri;
+
+      // Replace variables if needed
+      uri = StorageUtil.replaceMacro(checkNotNull(uri), run, listener);
+
+      // Ensure the uri starts with "gs://"
+      if (!uri.startsWith(GCS_SCHEME)) {
+        listener.error(prefix(moduleName,
+            Messages.AbstractUploadDescriptor_BadPrefix(
+                uri, GCS_SCHEME)));
+        bucket = "";
+        object = "";
+        return;
```

Comment:
Consider throwing an exception, rather than allowing this object to be initialized in an invalid state.

If you then let the caller handle logging and invoking replaceMacro, you could also eliminate all but the `uri` parameter and eliminate the alternate constructor below (the unit tests could just use this one).